### PR TITLE
Navigation: Fallback to a classic menu if one is available

### PR DESCRIPTION
**Pull request by scruffian**
## What?
If a site has a classic navigation menu, but no wp_navigation CPTs then we should fallback to the classic navigation block. If there's a primary location then we should use the navigation assigned to it, if not then we could use the latest one.

## Why?
For users who have set up classic menus this will be better than just showing them a list of pages.

## How?
In the frontend we just fetch the correct navigation menu and convert it to blocks

In the editor we actually convert the navigation menu to a wp_navigation CPT. This creates some issues - at the moment it's created as a draft. Also it gets created multiple times. Perhaps the best way round this is to create it as a published menu, so that the fallbacks will continue to work as they do right now.

## Testing Instructions
- You need to have a classic menu called primary and no wp_navigation CPTs
- In that scenario you should see the classic menu in both the editor and the frontend

### CODE REVIEW

File: packages/block-library/src/navigation/edit/index.js [hunk 1]
```diff
@@ -224,36 +224,54 @@ function Navigation( {
...
-		 *  The fallback should not request a save (entity dirty state)
-		 *  nor to be undoable, hence why it is marked as non persistent
-		 */
-		__unstableMarkNextChangeAsNotPersistent();
-		setRef( fallbackNavigationMenus[ 0 ].id );
+			// We should get the primary one.
+
+			// We should probably create this not as a draft...
+
+			// This is duplicated several times.
```
Comment (cbravobernal) [inline packages/block-library/src/navigation/edit/index.js]: Here we are going back to the problem of converting a classic menu to a block nav menu automatically on page load, if we do it as a draft, we will have duplicated draft menus. If we do it as published and the user reloads without saving, classic menu will be selected and shown in both editor and frontend.

Could be possible to add a nav menu from the classic menu on the editor without creating an entity in the database? Should we wait for slug references to add them as a check for not duplicating them?

Comment (talldan) [reply packages/block-library/src/navigation/edit/index.js]: > Could be possible to add a nav menu from the classic menu on the editor without creating an entity in the database? Should we wait for slug references to add them as a check for not duplicating them?

The blocks could be 'uncontrolled' inner blocks (so I think the code would get the converted blocks from the classic menu and set them using `replaceBlocks`), but they're not the smoothest experience either. 

The nice thing about creating an entity properly is that the same fallback logic won't need to be added in the PHP render code, so it's probably best either not to create it as a draft, or do the slug reference checking thing.

----

edit: just checked further down the PR, and I see that there's PHP render code anyway, I missed the fact that the block would need to be saved so that the menu could be shown on the frontend. 🤦

Comment (talldan) [reply packages/block-library/src/navigation/edit/index.js]: I guess the other option is that the creation of the wp_navigation happens server side, and then the client side fallback code that selects navigation menus would already handle the job on the client side.

There is a `gutenberg_convert_menu_items_to_blocks` PHP function already, but I don't know if it works perfectly. At a glance it doesn't look like it was ever updated to handle the navigation-submenu block.

**Comment by scruffian** [comment]
> If we do it as published and the user reloads without saving, classic menu will be selected and shown in both editor and frontend.

I don't think this is a problem, as the classic menu would already be showing in the frontend, as the conditions are the same:

1. If there's no wp_navigation menus, then show a classic menu on the front end
2. In the editor, if there's no wp_navigation menus then convert the classic menu to a wp_navigation menu
3. Now there is a wp_navigation menu...

**Comment by cbravobernal** [comment]
> For users who have set up classic menus this will be better than just showing them a list of pages.

I completely agree with this too 😄

**Comment by draganescu** [comment]
I think we should try and only address the most basic use case:

- website on classic theme with exactly one classic menu
- switches to block theme
- that one classic menu is automatically imported
- in this case the block menu logic works as the newly imported menu is the most recent one
- if the classic menu was already imported don't do nothing

If the website has multiple classic menus we should not do anything. We can come back later with some implementation that matches classic "location" to block template part "slug", but I think this would be more complex.

**Comment by scruffian** [comment]
I've made some changes to this. Now, if there's no wp_navigation CPTs and one classic menu, we convert the classic menu to a wp_navigation CPT, and this becomes the fallback. This happens in both the editor and the frontend.

File: packages/block-library/src/navigation/edit/index.js [hunk 2]
```diff
@@ -224,36 +224,54 @@ function Navigation( {
...
+					'publish'
+				);
+				if ( navMenu ) {
+					handleUpdateMenu( navMenu.id, {
+						focusNavigationBlock: true,
+					} );
+				}
+			};
+
+			onSelectClassicMenu( classicMenus[ 0 ] );
```
Comment (draganescu) [inline packages/block-library/src/navigation/edit/index.js]: I think here we should use the same process that we use for the block menus, that is avoid calling `handleUpdateMenu`. This call sets the ref and the use sees a dirty post. We should instead just _create_ the block menu, then _because it's the latest created_ it will be handled by the later code.

File: packages/block-library/src/navigation/edit/index.js [hunk 3]
```diff
@@ -640,7 +658,8 @@ function Navigation( {
 							onSelectClassicMenu={ async ( classicMenu ) => {
 								const navMenu = await convertClassicMenu(
 									classicMenu.id,
-									classicMenu.name
+									classicMenu.name,
+									'draft'
```
Comment (draganescu) [inline packages/block-library/src/navigation/edit/index.js:665]: What is this change for?

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js [hunk 4]
```diff
@@ -32,7 +32,11 @@ function useConvertClassicToBlockMenu( clientId ) {
 	const [ status, setStatus ] = useState( CLASSIC_MENU_CONVERSION_IDLE );
 	const [ error, setError ] = useState( null );
 
-	async function convertClassicMenuToBlockMenu( menuId, menuName ) {
+	async function convertClassicMenuToBlockMenu(
+		menuId,
+		menuName,
+		postStatus = 'publish'
```
Comment (draganescu) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:42]: I left a comment in the edit component as well, I may be missing why we would import as draft. The change to this function is not bad, it's fine to be able to specify the status on import, but why did we change the importing status on the nav selector?

File: packages/block-library/src/navigation/index.php [hunk 5]
```diff
@@ -248,13 +248,63 @@ function block_core_navigation_render_submenu_icon() {
...
+	unset( $menu_items, $menu_item );
+
+	// END: Code that already exists in wp_nav_menu().
+
+	$menu_items_by_parent_id = array();
+	foreach ( $sorted_menu_items as $menu_item ) {
+		$menu_items_by_parent_id[ $menu_item->menu_item_parent ][] = $menu_item;
+	}
+
+	$inner_blocks = gutenberg_convert_menu_items_to_blocks(
```
Comment (draganescu) [inline packages/block-library/src/navigation/index.php]: We'll need to update this function to use `Submenu` instead of nesting link blocks into link blocks. Ideally we could do it in a separate PR?

File: packages/block-library/src/navigation/index.php [hunk 6]
```diff
@@ -322,6 +372,29 @@ function block_core_navigation_get_fallback_blocks() {
 
 	$navigation_post = block_core_navigation_get_most_recently_published_navigation();
 
+	// If there are no navigation posts then try to find a classic menu.
```
Comment (draganescu) [inline packages/block-library/src/navigation/index.php]: I would:
- make this whole new block a new function :) 
- put the contents of `block_core_navigation_get_classic_menu_fallback` into this new function (unlikely we'd reuse that)
- call this new function in `block_core_navigation_get_most_recently_published_navigation` as a case so that here we only call it once.

**Review by draganescu — CHANGES_REQUESTED** [review]
Awesome! The code looks almost there, I left some tweaks. Requesting changes mostly about making sure we maintain the fallback behavior of not dirtying the post on fallback.

Did not test visually yet.

Later edit:, visual testing results:

- [x] The template part ends up dirty upon fallback. As mentioned in code, we should not assign the ref, but allow the fallback to set it marked as  `__unstableMarkNextChangeAsNotPersistent`.
- [x] The import happens multiple times:
<details>
<summary>Screenshot - Initial load</summary>
image
</details>
<details>
<summary>Screenshot - Second refresh</summary>
image
</details>

- [x] The menu is imported as `draft`. Considering this is the only classic menu that exists it's fine to import it as `published`.
<details>
<summary>All menus are draft</summary>
image

</details>

File: packages/block-library/src/navigation/edit/index.js (hunk 3 above)
Comment (scruffian) [reply packages/block-library/src/navigation/edit/index.js:665]: When we import classic menus usually, we set them to draft so that the fallback behaviour isn't changed until the draft is published (which happens when the user clicks save). In this case, we want the classic menu to be published straight away as it is already behaving as the fallback on the frontend.

Comment (draganescu) [reply packages/block-library/src/navigation/edit/index.js:665]: Aha! Got this.

File: packages/block-library/src/navigation/index.php (hunk 6 above)
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: I think we'll always need to call `block_core_navigation_get_most_recently_published_navigation` twice - once so we can determine whether there are any wp_navigation CPTs and then, if there aren't any but there is a classic navigation menu, to get the newly created one from the import.

Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: I did a bit of a refactor here nonetheless...

File: packages/block-library/src/navigation/index.php (hunk 5 above)
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: I did it in this PR but I can break it out into a different on if you think it's worth it.

**Comment by scruffian** [comment]
Thanks for testing @draganescu, I have addressed all the feedback here, so it's ready for another look.

**Review by draganescu — APPROVED** [review]
Tested this again after [e5f8fb5](https://github.com/WordPress/gutenberg/pull/44173/commits/e5f8fb590143aa9224dcac40c7b4b426d7c7c801) and it worked pretty smooth:

- one classic menu resulted in auto importing it and the block theme's empty navigation block used it as a fallback.

File: lib/experimental/navigation-theme-opt-in.php [hunk 7]
```diff
@@ -191,36 +191,33 @@ function gutenberg_convert_menu_items_to_blocks(
...
-			$menu_items_by_parent_id
+		$block = array(
+			'blockName' => $block_name,
+			'attrs'     => array(
+				'label' => $menu_item->title,
+				'url'   => $menu_item->url,
+			),
 		);
-
+		$block_content = gutenberg_convert_menu_items_to_blocks(
```
Comment (draganescu) [inline lib/experimental/navigation-theme-opt-in.php]: I don't think we can use this? This is in the experimental part of the defunct navigation editor, I think we have `block_core_navigation_parse_blocks_from_menu_items` for this use which is already used for `__unstableLocation``

**Review by draganescu — DISMISSED** [review]

File: packages/block-library/src/navigation/edit/index.js [hunk 8]
```diff
@@ -224,36 +224,54 @@ function Navigation( {
...
-			return menuADate.getTime() < menuBDate.getTime();
-		} );
+		// If there's non fallback navigation menus and
+		// only one classic menu then create a new navigation menu based on it.
+		if ( ! fallbackNavigationMenus?.length && classicMenus?.length === 1 ) {
+			// Only create classic menus once.
+			if ( ! convertingClassicMenu ) {
+				// This is duplicated several times.
+				const onSelectClassicMenu = async ( classicMenu ) => {
+					setConvertingClassicMenu( true );
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]: I wonder if we could avoid this and use [the conversion status data already exposed by the classic menu hook](https://github.com/WordPress/gutenberg/blob/349240ac3ea4890f0a95b07e4d089b4307a6eab7/packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js#L149).

File: packages/block-library/src/navigation/edit/index.js [hunk 9]
```diff
@@ -339,6 +357,7 @@ function Navigation( {
 	// Turn on contrast checker for web only since it's not supported on mobile yet.
 	const enableContrastChecking = Platform.OS === 'web';
 
+	const [ convertingClassicMenu, setConvertingClassicMenu ] = useState();
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]: See previous comment about already having [this state available from the hoo](https://github.com/WordPress/gutenberg/blob/349240ac3ea4890f0a95b07e4d089b4307a6eab7/packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js#L149)k.

File: packages/block-library/src/navigation/edit/index.js (hunk 3 above)
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js:665]: > In this case, we want the classic menu to be published straight away as it is already behaving as the fallback on the frontend.

I'm confused about this. The code is showing the menu being created as a `draft` status.

File: packages/block-library/src/navigation/index.php [hunk 10]
```diff
@@ -248,13 +248,93 @@ function block_core_navigation_render_submenu_icon() {
...
+ * Get the classic navigation menu to use as a fallback.
+ *
+ * @return object WP_Term The classic navigation.
+ */
+function block_core_navigation_get_classic_menu_fallback() {
+	$classic_nav_menus = wp_get_nav_menus();
+
+	// If menus exist.
+	if ( $classic_nav_menus && ! is_wp_error( $classic_nav_menus ) && count( $classic_nav_menus ) === 1 ) {
+		// If there's only one classic menu then use it.
```
Comment (getdave) [inline packages/block-library/src/navigation/index.php]:
```suggestion
		// Use the first classic menu only. Handles simple use case where user has a single
		// classic menu and switches to a block theme. In future this maybe expanded to
		// determine the most appropriate classic menu to be used based on location.
```

Just wondering whether we should add more context here.

File: packages/block-library/src/navigation/index.php [hunk 11]
```diff
@@ -248,13 +248,93 @@ function block_core_navigation_render_submenu_icon() {
...
+	_wp_menu_item_classes_by_context( $menu_items );
+
+	$sorted_menu_items = array();
+	foreach ( (array) $menu_items as $menu_item ) {
+		$sorted_menu_items[ $menu_item->menu_order ] = $menu_item;
+	}
+
+	unset( $menu_items, $menu_item );
+
+	// END: Code that already exists in wp_nav_menu().
```
Comment (getdave) [inline packages/block-library/src/navigation/index.php:288]: Where does this `START:`?

File: packages/block-library/src/navigation/index.php [hunk 12]
```diff
@@ -248,13 +248,93 @@ function block_core_navigation_render_submenu_icon() {
...
+	// See if we have a classic menu.
+	$classic_nav_menu = block_core_navigation_get_classic_menu_fallback();
+
+	// If we have a classic menu then convert it to blocks.
+	if ( $classic_nav_menu ) {
+		$classic_nav_menu_blocks            = block_core_navigation_get_classic_menu_fallback_blocks( $classic_nav_menu );
+		$classic_nav_menu_blocks_serialized = serialize_blocks( $classic_nav_menu_blocks );
+
+		// Create a new navigation menu from the classic menu.
+		wp_insert_post(
```
Comment (getdave) [inline packages/block-library/src/navigation/index.php]: Suggestion: this will give us an ID. From that we can [`get_post()`](https://developer.wordpress.org/reference/functions/get_post/). That will _ensure_ we get the exact same post that was created.

File: packages/block-library/src/navigation/index.php [hunk 13]
```diff
@@ -248,13 +248,93 @@ function block_core_navigation_render_submenu_icon() {
...
+}
 
 /**
  * Finds the most recently published `wp_navigation` Post.
  *
  * @return WP_Post|null the first non-empty Navigation or null.
  */
 function block_core_navigation_get_most_recently_published_navigation() {
+
 	// We default to the most recently created menu.
```
Comment (getdave) [inline packages/block-library/src/navigation/index.php]:
```suggestion
	// Default to the most recently created menu.
```

File: packages/block-library/src/navigation/index.php (hunk 12 above)
Comment (spacedmonkey) [reply packages/block-library/src/navigation/index.php]: wp_insert_post can return an error. This function, doesn't always, "just work". Where is the error handling here?

Comment (getdave) [reply packages/block-library/src/navigation/index.php]: For reference https://developer.wordpress.org/reference/functions/wp_insert_post/#return

File: packages/block-library/src/navigation/index.php [hunk 14]
```diff
@@ -248,14 +248,96 @@ function block_core_navigation_render_submenu_icon() {
...
+	// If we have a classic menu then convert it to blocks.
+	if ( $classic_nav_menu ) {
+		$classic_nav_menu_blocks            = block_core_navigation_get_classic_menu_fallback_blocks( $classic_nav_menu );
+		$classic_nav_menu_blocks_serialized = serialize_blocks( $classic_nav_menu_blocks );
+
+		// Create a new navigation menu from the classic menu.
+		wp_insert_post(
+			array(
+				'post_content' => $classic_nav_menu_blocks_serialized,
+				'post_title'   => $classic_nav_menu->slug,
```
Comment (spacedmonkey) [inline packages/block-library/src/navigation/index.php]:
```suggestion
				'post_title'   => $classic_nav_menu->slug,
				'post_name'   => $classic_nav_menu->slug,
```

File: packages/block-library/src/navigation/edit/index.js (hunk 3 above)
Comment (scruffian) [reply packages/block-library/src/navigation/edit/index.js:665]: Right, this is inside the part of the code that handles the conversion on demand. In this case we want it to be draft. However if we are doing an automatic conversion we want that one to be published.

File: packages/block-library/src/navigation/index.php (hunk 12 above)
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: In the future we may not want to get that menu, if the logic for the fallbacks changes. By just calling `block_core_navigation_get_most_recently_published_navigation` again we are letting that function be the authority on what the correct fallback should be, so that if we update the behaviour we won't need to also update the logic here.

Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: I've added error handling for the wp_insert_post call.

File: lib/experimental/navigation-theme-opt-in.php (hunk 7 above)
Comment (scruffian) [reply lib/experimental/navigation-theme-opt-in.php]: Updated to use this function

File: packages/block-library/src/navigation/index.php [hunk 15]
```diff
@@ -323,6 +323,10 @@ function block_core_navigation_maybe_use_classic_menu_fallback() {
 			)
 		);
 
+		if ( is_wp_error( $wp_insert_post_result ) ) {
```
Comment (spacedmonkey) [inline packages/block-library/src/navigation/index.php]: To do this, you need to pass, second param of `wp_insert_post` as true.

Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: Thanks! Done

File: packages/block-library/src/navigation/edit/index.js [hunk 16]
```diff
@@ -224,36 +224,54 @@ function Navigation( {
...
-			! navigationMenus?.length
-		) {
+		// Only autofallback to published menus.
+		const fallbackNavigationMenus = navigationMenus?.filter(
+			( menu ) => menu.status === 'publish'
+		);
+
+		if ( hasUncontrolledInnerBlocks || isCreatingNavigationMenu || ref ) {
 			return;
 		}
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]:
```suggestion
		if ( hasUncontrolledInnerBlocks || isCreatingNavigationMenu || ref ) {
			return;
		}
		
		// Only autofallback to published menus.
		const fallbackNavigationMenus = navigationMenus?.filter(
			( menu ) => menu.status === 'publish'
		);
```

Let's exit early.

File: packages/block-library/src/navigation/edit/index.js (hunk 8 above)
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: Fixed in https://github.com/WordPress/gutenberg/pull/44173/commits/1c8f006cc099283a6f32875f38948f9a9685aa74

File: packages/block-library/src/navigation/edit/index.js (hunk 9 above)
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: I've removed this state and used the existing state in https://github.com/WordPress/gutenberg/pull/44173/commits/1c8f006cc099283a6f32875f38948f9a9685aa74

File: packages/block-library/src/navigation/edit/index.js [hunk 17]
```diff
@@ -256,16 +264,26 @@ function Navigation( {
...
-		status: classicMenuConversionStatus,
-		error: classicMenuConversionError,
-	} = useConvertClassicToBlockMenu( clientId );
+		// If there's non fallback navigation menus and
+		// only one classic menu then create a new navigation menu based on it.
+		await convertClassicMenu(
+			classicMenus[ 0 ].id,
+			classicMenus[ 0 ].name,
+			'publish'
+		);
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]:
```suggestion
		await convertClassicMenu(
			classicMenus[ 0 ].id,
			classicMenus[ 0 ].name,
			'publish'
		).catch( () => {} );
```

We're not catching potential errors here. Ditto all other cases where we are `await`ing the result of the function.

File: packages/block-library/src/navigation/edit/index.js [hunk 18]
```diff
@@ -280,9 +298,10 @@ function Navigation( {
 		! isCreatingNavigationMenu &&
 		! isConvertingClassicMenu &&
 		hasResolvedNavigationMenus &&
+		classicMenus?.length === 0 &&
 		! hasUncontrolledInnerBlocks;
 
-	if ( isPlaceholder && ! ref ) {
+	if ( isPlaceholder ) {
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]: We should probably explain why this was removed.

File: packages/block-library/src/navigation/index.php [hunk 19]
```diff
@@ -248,14 +248,102 @@ function block_core_navigation_render_submenu_icon() {
...
+ * If there's a the classic menu then use it as a fallback.
+ *
+ * @return array the normalized parsed blocks.
+ */
+function block_core_navigation_maybe_use_classic_menu_fallback() {
+	// See if we have a classic menu.
+	$classic_nav_menu = block_core_navigation_get_classic_menu_fallback();
+
+	// If we have a classic menu then convert it to blocks.
+	if ( $classic_nav_menu ) {
```
Comment (getdave) [inline packages/block-library/src/navigation/index.php]: Rather than nesting can we just invert this and return early?

File: packages/block-library/src/navigation/index.php [hunk 20]
```diff
@@ -248,14 +248,102 @@ function block_core_navigation_render_submenu_icon() {
...
+		$classic_nav_menu_blocks            = block_core_navigation_get_classic_menu_fallback_blocks( $classic_nav_menu );
+		$classic_nav_menu_blocks_serialized = serialize_blocks( $classic_nav_menu_blocks );
+
+		// Create a new navigation menu from the classic menu.
+		$wp_insert_post_result = wp_insert_post(
+			array(
+				'post_content' => $classic_nav_menu_blocks_serialized,
```
Comment (getdave) [inline packages/block-library/src/navigation/index.php]: The classic menu may be empty. Do we want to use it as a fallback in this case?

File: packages/block-library/src/navigation/index.php [hunk 21]
```diff
@@ -322,6 +410,11 @@ function block_core_navigation_get_fallback_blocks() {
 
 	$navigation_post = block_core_navigation_get_most_recently_published_navigation();
 
+	// If there are no navigation posts then try to find a classic menu.
```
Comment (getdave) [inline packages/block-library/src/navigation/index.php]:
```suggestion
	// If there are no navigation posts then try to find a classic menu
	// and convert it into a block based navigation menu.
```

File: packages/block-library/src/navigation/index.php [hunk 22]
```diff
@@ -322,6 +410,11 @@ function block_core_navigation_get_fallback_blocks() {
...
+	if ( ! $navigation_post ) {
+		$navigation_post = block_core_navigation_maybe_use_classic_menu_fallback();
+	}
+
 	// Prefer using the first non-empty Navigation as fallback if available.
```
Comment (getdave) [inline packages/block-library/src/navigation/index.php]:
```suggestion
	// Use the first non-empty Navigation as fallback if available.
```

File: packages/block-library/src/navigation/edit/index.js [hunk 23]
```diff
@@ -256,16 +264,26 @@ function Navigation( {
 		setRef( fallbackNavigationMenus[ 0 ].id );
 	}, [ navigationMenus ] );
 
-	const navRef = useRef();
+	useEffect( async () => {
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]: As I mentioned during our pairing call on this feature, using `async` in an effect is not best practice. Why?

Because the effect is supposed to return either nothing or a cleanup _function_, However `async` will always return a `Promise` which is not a function. 

While this doesn't break in React17, it's not a good practice and it will start breaking in React 18.

An alternative is something like

```js
useEffect(() => {
    (async () => {
        // do stuff
    })();
  }, []);
```


Here is [a previous PR which had to clear up the same problem](https://github.com/WordPress/gutenberg/pull/40373).

File: packages/block-library/src/navigation/edit/index.js (hunk 18 above)
Comment (scruffian) [reply packages/block-library/src/navigation/edit/index.js]: we didn't remove it, it's already being checked as part of `isPlaceholder`.

Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: 👍

**Comment by scruffian** [comment]
I found a bug with this. When you have two navigation blocks on the page they will both try to create a classic navigation, and do so successfully.

File: lib/experimental/navigation-theme-opt-in.php (hunk 7 above)
Comment (draganescu) [reply lib/experimental/navigation-theme-opt-in.php]: We should then undo these changes?

**Comment by scruffian** [comment]
@draganescu and I created a solution to prevent menus from importing twice. We'd be grateful for a review.

We tried creating a store and then saving state of the import in the store, but the store didn't seem to update fast enough so we still experienced the problem of the import happening twice. This module level variable seems to solve it though...

File: lib/experimental/navigation-theme-opt-in.php (hunk 7 above)
Comment (scruffian) [reply lib/experimental/navigation-theme-opt-in.php]: Done

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js [hunk 24]
```diff
@@ -111,7 +120,15 @@ function useConvertClassicToBlockMenu( clientId ) {
 		return navigationMenu;
 	}
 
-	const convert = useCallback( async ( menuId, menuName ) => {
+	const convert = useCallback( async ( menuId, menuName, postStatus ) => {
+		// Check whether this classic menu is being imported already.
+		if ( isConvertingClassicMenu === menuId ) {
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]:
```suggestion
		if ( classicMenuBeingConvertedId === menuId ) {
```

Nit 🤷‍♂️

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js [hunk 25]
```diff
@@ -17,6 +17,10 @@ export const CLASSIC_MENU_CONVERSION_ERROR = 'error';
 export const CLASSIC_MENU_CONVERSION_PENDING = 'pending';
 export const CLASSIC_MENU_CONVERSION_IDLE = 'idle';
 
+// This is needed to ensure that multiple components using this hook
+// do not import the same classic menu twice.
+let isConvertingClassicMenu = null;
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]:
```suggestion
let classicMenuBeingConvertedId = null;
```

Nit: 🤷‍♂️

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js [hunk 26]
```diff
@@ -111,7 +120,15 @@ function useConvertClassicToBlockMenu( clientId ) {
...
 	}
 
-	const convert = useCallback( async ( menuId, menuName ) => {
+	const convert = useCallback( async ( menuId, menuName, postStatus ) => {
+		// Check whether this classic menu is being imported already.
+		if ( isConvertingClassicMenu === menuId ) {
+			return;
+		}
+
+		// Set the ID for the currently importing classic menu.
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:129]: Nit: I'm not sure this comment adds a tonne of value

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js [hunk 27]
```diff
@@ -121,13 +138,20 @@ function useConvertClassicToBlockMenu( clientId ) {
...
+		)
 			.then( ( navigationMenu ) => {
 				setStatus( CLASSIC_MENU_CONVERSION_SUCCESS );
+				// Reset the ID for the currently importing classic menu.
+				isConvertingClassicMenu = null;
 				return navigationMenu;
 			} )
 			.catch( ( err ) => {
 				setError( err?.message );
+				// Reset the ID for the currently importing classic menu.
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:154]:
```suggestion
				// Reset the ID for the currently importing classic menu.
				isConvertingClassicMenu = null;
```

I assume we want to reset the menu being converted if it failed to convert?

File: packages/block-library/src/navigation/index.php [hunk 28]
```diff
@@ -248,14 +248,109 @@ function block_core_navigation_render_submenu_icon() {
...
+	// See if we have a classic menu.
+	$classic_nav_menu = block_core_navigation_get_classic_menu_fallback();
+
+	if ( ! $classic_nav_menu ) {
+		return;
+	}
+
+	// If we have a classic menu then convert it to blocks.
+	$classic_nav_menu_blocks            = block_core_navigation_get_classic_menu_fallback_blocks( $classic_nav_menu );
+	$classic_nav_menu_blocks_serialized = serialize_blocks( $classic_nav_menu_blocks );
```
Comment (getdave) [inline packages/block-library/src/navigation/index.php]: This feels like an implementation detail that should be encapsulated within `block_core_navigation_get_classic_menu_fallback_blocks()`.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js [hunk 29]
```diff
@@ -17,6 +17,10 @@ export const CLASSIC_MENU_CONVERSION_ERROR = 'error';
 export const CLASSIC_MENU_CONVERSION_PENDING = 'pending';
 export const CLASSIC_MENU_CONVERSION_IDLE = 'idle';
 
+// This is needed to ensure that multiple components using this hook
+// do not import the same classic menu twice.
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:21]: This comment is helpful 👍

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js (hunk 27 above)
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:154]: Going to test this

Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:154]: Confirmed fixed.

File: packages/block-library/src/navigation/index.php (hunk 28 above)
Comment (scruffian) [reply packages/block-library/src/navigation/index.php]: Done

**Review by draganescu — APPROVED** [review]
Tested and did what it says. The code is pretty streamlined and LGTM.

**Comment by getdave** [comment]
@draganescu Just confirming that this _won't_ now make it into 6.1 right?

**Comment by scruffian** [comment]
I believe that is the case

**Comment by draganescu** [comment]
I think it would benefit from waiting in the plugin more. What navigation building experience gained for 6.1 is definitely good, this would work best once the other work on slug based menu migration is also complete.
